### DISCUSSION TITLE
Is it a good practise to update the coordinates and recalculate the spherical harmonics at each layer?

### DISCUSSION THREAD

**Discussion by haotianteng** [e3nn/e3nn#439] [General]
Hi,
I am wondering if it's meaningful to update the coordinates after each layer and recalculate the spherical harmonics function with the updated coordinates.

In the example code, the tensor product only updates the node attribute with constant spherical harmonics calculated from the original coordinates.

```
edge_features = tp1(node_features, **edge_sh**)
node_features = scatter(edge_features, dim=0).div(num_neighbors**0.5)
edge_features = tp2(node_features, **edge_sh**)
node_features = scatter(edge_features, dim=0).div(num_neighbors**0.5)
```

Where `edge_sh` usually does not change over different tensor product operations and layers, are the following operations valid to perform? (i.e., do they break the equivariance of the network):

```
edge_features = tp1(node_features, **edge_sh**)
node_features = scatter(edge_features, dim=0).div(num_neighbors**0.5)

# Update the coordinates
pos += node_features[:3] #assume the output of tp1 is "1x1e"
edge_vec = pos[src] - pos[dest]
edge_sh = o3.spherical_harmonics(l = o3.Irreps.spherical_harmonics(3), edge_vec)

edge_features = tp2(node_features, **edge_sh**)
node_features = scatter(edge_features, dim=0).div(num_neighbors**0.5)

```

**Comment by mariogeiger** [comment]
Looks good to me. The spherical harmonics are relatively cheap to compute compare to the rest of the model anyway

**Comment by haotianteng** [comment]
Do I need to use an odd irrpes (1o) if I want to add it to the coordinate (as coordinate vector is not a pseudovector)?
Also is there a way to transfer back from a sh irreps to the original vector? Such as an L up to 3 irreps back to its edge vector, I feel adding this restored edge vector back to the coordinates makes more sense, as it's now in coordinate space.

**Reply by mariogeiger** [reply]
Yes you must add 1o to coordinates to respect O3 equivariance. 

Spherical harmonics Y2 is a degree 2 polynomial, Y3 is a degree 3 polynomial. Etc

These polynomials are injective but not surjective. Inverting them is non trivial. I tried to invert Y2 for a project but I have no idea how to do that for Y3. I would just create 1o with (learned) polynomials

**Reply by haotianteng** [reply]
Thanks for the clarification.
